File: ruby/TensionsBetweenDifferentOOPrinciples/TensionBetweenDRYandLowCoupling/After/rosie_robot.rb
require_relative 'robot'

module TensionBetweenDRYandLowCoupling
  module After
    class RosieRobot < Robot
      # RosieRobot class represents a robot that accept commands to move in four directions.
      # RosieRobot class follows the DRY principle now because code in common with UniblabRobot is in Robot class.
      # RosieRobot class has one more dependency to Robot class so coupling is higher.
 
      def initialize
        @current_position = [5, 5]
      end
      
      def move_one_north
        @current_position = add @current_position, North
      end

      def move_one_south
        @current_position = add @current_position, South
      end

      def move_one_east
        @current_position = add @current_position, East
      end

      def move_one_west
        @current_position = add @current_position, West
      end
    end
  end
end

File: Java/tensionsBetweenDifferentOOPrinciples/src/main/java/tensionBetweenDRYandLowCoupling/after/RosieRobot.java

package tensionsBetweenDifferentOOPrinciples.tensionBetweenDRYandLowCoupling.after;

import tensionsBetweenDifferentOOPrinciples.Tuple;

public class RosieRobot extends Robot {
    // RosieRobot class represents a robot that accept commands to move in four directions.
    // RosieRobot class follows the DRY principle now because code in common with UniblabRobot is in Robot class.
    // RosieRobot class has one more dependency to Robot class so coupling is higher.

    private Tuple<Integer, Integer> _currentPosition = new Tuple<Integer, Integer>(5, 5);

    public void MoveOneNorth() {
        _currentPosition = Add(_currentPosition, North);
    }

    public void MoveOneSouth() {
        _currentPosition = Add(_currentPosition, South);
    }

    public void MoveOneEast() {
        _currentPosition = Add(_currentPosition, East);
    }

    public void MoveOneWest() {
        _currentPosition = Add(_currentPosition, West);
    }
}

